orttraining/orttraining/training_ops/cuda/communication/common.h

```diff
@@ -26,5 +27,79 @@ inline size_t GetAggregatedAlignedAddress(size_t old_addr) {
   return new_addr;
 }
 
+// This function extracts shapes and sizes for tensors indexed by
+// begin, begin + 1, ..., begin + count - 1.
+// tensor_sizes[i]: i-th indexed tensor's size.
+// tensor_shapes[i]: i-th indexed tensor's shape.
+inline void GetTensorShapesAndSizes(
+    const bool is_index_input,                  // Index inputs from the context if true. Otherwise, use outputs.
+    const int begin,                            // Index of the first sent/received tensor in the context.
+    const int count,                            // Number of sent/received tensors.
+    OpKernelContext* ctx,                       // The context.
+    std::vector<size_t>& tensor_sizes,          // tensor_sizes[i] is the size of i-th sent/received tensor in byte.
+    std::vector<TensorShape>& tensor_shapes) {  // tensor_shapes[i] is the i-th sent/received tensor.
+
+  // Helper function to retrieve input or output tensor.
+  auto get_tensor = [&](const int index) -> const Tensor* {
+    if (is_index_input) {
+      return ctx->Input<Tensor>(begin + index);
+    } else {
+      return ctx->Output<Tensor>(begin + index);
+    }
+  };
+
+  // Get tensors and shapes for indexed tensors from context.
+  tensor_sizes.resize(count);
+  tensor_shapes.resize(count);
+  for (int i = 0; i < count; ++i) {
+    const Tensor* tensor = get_tensor(i);
+    tensor_sizes[i] = tensor->SizeInBytes();
```

Comment:
>tensor_sizes [](start = 4, length = 12)

What do we do for tensor_sizes[i] when i is less than begin? If we don't need them, maybe we don't need to reserve slot in the vector for those data?

Reply:
We don't reserve slots for tensor before `begin`. Here `i=0` maps the `input[begin]` (see `get_tensor` function)

---
In reply to: [459819913](https://github.com/microsoft/onnxruntime/pull/4547#discussion_r459819913) [](ancestors = 459819913)

Reply:
Ah, I thought the `count` in parameter is the count for all inputs/outputs which includes the offset from `begin`. It makes sense it is the count of `element_types`